### PR TITLE
Define state-neutral events

### PR DESCRIPTION
**Pull request by AlexanderLitus**
Depends on PR #9.

### CODE REVIEW

File: registration/src/main/java/org/spine3/samples/lobby/registration/order/OrderAggregate.java [hunk 1]
```diff
@@ -247,22 +238,13 @@ public boolean apply(@Nullable SeatQuantity requestedOne) {
...
-                final boolean result =
-                        (seat != null) &&
-                        seat.hasSeatTypeId() &&
-                        seat.getSeatTypeId().equals(id);
-                return result;
-            }
-        }, defaultValue);
-        return result;
+    @Override
+    @SuppressWarnings({"RefusedBequest", "ReturnOfCollectionOrArrayField"/**it is immutable*/})
```
Comment (alexander-yevsyukov) [inline registration/src/main/java/org/spine3/samples/lobby/registration/order/OrderAggregate.java:242]: Do we really need to have two leading stars in the comment?

File: registration/src/main/java/org/spine3/samples/lobby/registration/seat/availability/SeatsAvailabilityAggregate.java [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+
+        final SeatsReservationCommitted.Builder event = SeatsReservationCommitted.newBuilder()
+                .setReservationId(cmd.getReservationId());
+        return event.build();
+    }
+
+    @Assign
+    public SeatsReservationCancelled handle(CancelSeatReservation cmd, CommandContext context) {
+        Validator.validateCommand(cmd);
+        Validator.validateState(getState(), cmd);
```
Comment (alexander-yevsyukov) [inline registration/src/main/java/org/spine3/samples/lobby/registration/seat/availability/SeatsAvailabilityAggregate.java:97]: Extract state into local var.

File: registration/src/main/java/org/spine3/samples/lobby/registration/seat/availability/SeatsAvailabilityAggregate.java [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+    }
+
+    @Override
+    protected SeatsAvailability getDefaultState() {
+        return SeatsAvailability.getDefaultInstance();
+    }
+
+    @Assign
+    public SeatsReserved handle(MakeSeatReservation cmd, CommandContext context) {
+        Validator.validateCommand(cmd);
```
Comment (alexander-yevsyukov) [inline registration/src/main/java/org/spine3/samples/lobby/registration/seat/availability/SeatsAvailabilityAggregate.java:71]: This construct is mouthful. It should be just `validateCommand()` and `validateState()`.

File: registration/src/main/java/org/spine3/samples/lobby/registration/seat/availability/SeatsAvailabilityAggregate.java [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+    @Assign
+    public RemovedAvailableSeats handle(RemoveSeats cmd, CommandContext context) {
+        Validator.validateCommand(cmd);
+        Validator.validateState(getState(), cmd);
+
+        final RemovedAvailableSeats.Builder event = RemovedAvailableSeats.newBuilder()
+                .setQuantity(cmd.getQuantity());
+        return event.build();
+    }
+
```
Comment (alexander-yevsyukov) [inline registration/src/main/java/org/spine3/samples/lobby/registration/seat/availability/SeatsAvailabilityAggregate.java:130]: Please have a section break here.

```
   /* Event Appliers */
```

File: registration/src/main/java/org/spine3/samples/lobby/registration/seat/availability/SeatsAvailabilityAggregate.java [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+        if (newQuantity < 0) {
+            return 0;
+        }
+        return newQuantity;
+    }
+
+    /**
+     * The Method Object for handling {@link MakeSeatReservation} commands.
+     */
+    protected static class MakeSeatReservationCommandHandler {
```
Comment (alexander-yevsyukov) [inline registration/src/main/java/org/spine3/samples/lobby/registration/seat/availability/SeatsAvailabilityAggregate.java:215]: I've move this class to the level of the package, together with `Validator`. This aggregate is getting too big already.

File: registration/src/main/java/org/spine3/samples/lobby/registration/util/CollectionUtils.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+import static com.google.common.collect.Iterables.find;
+
+/**
+ * The utility class for working with collections.
+ *
+ * @author Alexander Litus
+ */
+@SuppressWarnings("UtilityClass")
+public class CollectionUtils {
```
Comment (alexander-yevsyukov) [inline registration/src/main/java/org/spine3/samples/lobby/registration/util/CollectionUtils.java:37]: Can these go to `Seats` class? See my comments to #9.

These are not generic collection utils.

File: registration/src/main/java/org/spine3/samples/lobby/registration/util/ValidationUtils.java [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+        checkMessageField(!seats.isEmpty(), "seats", message);
+        for (SeatQuantity seat : seats) {
+            checkMessageField(seat.hasSeatTypeId(), "seat type ID", seat);
+            if (seat.getQuantity() < 0) {
+                throw new IllegalArgumentException("The seat quantity must be positive.");
+            }
+        }
+    }
+
+    /**
```
Comment (alexander-yevsyukov) [inline registration/src/main/java/org/spine3/samples/lobby/registration/util/ValidationUtils.java:91]: This should go into Spine core.

File: registration/src/main/java/org/spine3/samples/lobby/registration/util/ValidationUtils.java [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+/**
+ * The utility class for validating messages (commands, events etc).
+ *
+ * @author Alexander Litus
+ */
+@SuppressWarnings("UtilityClass")
+public class ValidationUtils {
+
+    private ValidationUtils() {
```
Comment (alexander-yevsyukov) [inline registration/src/main/java/org/spine3/samples/lobby/registration/util/ValidationUtils.java:38]: Can we move these check methods into corresponding utility classes?

File: registration/src/main/java/org/spine3/samples/lobby/registration/seat/availability/SeatsAvailabilityAggregate.java (hunk 3 above)
Comment (AlexanderLitus) [inline registration/src/main/java/org/spine3/samples/lobby/registration/seat/availability/SeatsAvailabilityAggregate.java:71]: Moved to the package level and used static import.

Comment (alexander-yevsyukov) [inline registration/src/main/java/org/spine3/samples/lobby/registration/seat/availability/SeatsAvailabilityAggregate.java:71]: Move the Validator class, then static import.

**Pull request merged**